### PR TITLE
ARROW-11660: [C++] Move RecordBatch::SelectColumns method from R to C++ library

### PR DESCRIPTION
**Pull request by thisisnic**

### CODE REVIEW

File: cpp/src/arrow/record_batch.cc [hunk 1]
```diff
@@ -253,6 +253,26 @@ bool RecordBatch::ApproxEquals(const RecordBatch& other) const {
...
+  for (int i = 0; i < n; i++) {
+    int pos = indices[i];
+    if (pos < 0 || pos > num_columns() - 1) {
+      return Status::Invalid("Invalid column index ", pos, " to select columns.");
+    }
+    fields[i] = schema()->field(pos);
+    columns[i] = column(pos);
+  }
+  
+  auto schema = std::make_shared<arrow::Schema>(std::move(fields));
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/record_batch.cc:271]: Here we probably want to preserve the metadata of the schema (see the `Table::SelectColumns` method)

Comment (thisisnic) [reply cpp/src/arrow/record_batch.cc:271]: Oops, completely missed that - will update now, thanks for highlighting!

File: cpp/src/arrow/record_batch.cc [hunk 2]
```diff
@@ -253,6 +253,27 @@ bool RecordBatch::ApproxEquals(const RecordBatch& other) const {
   return true;
 }
 
+Result<std::shared_ptr<RecordBatch>> RecordBatch::SelectColumns(
+    const std::vector<int>& indices) const {
+  int n = static_cast<int>(indices.size());
+
+  std::vector<std::shared_ptr<Field>> fields(n);
+  std::vector<std::shared_ptr<Array>> columns(n);
```
Comment (pitrou) [inline cpp/src/arrow/record_batch.cc:261]: Nit, but you can use the `FieldVector` and `ArrayVector` type aliases, respectively.

File: cpp/src/arrow/record_batch.cc [hunk 3]
```diff
@@ -253,6 +253,27 @@ bool RecordBatch::ApproxEquals(const RecordBatch& other) const {
...
+    if (pos < 0 || pos > num_columns() - 1) {
+      return Status::Invalid("Invalid column index ", pos, " to select columns.");
+    }
+    fields[i] = schema()->field(pos);
+    columns[i] = column(pos);
+  }
+
+  auto new_schema =
+      std::make_shared<arrow::Schema>(std::move(fields), schema()->metadata());
+  return RecordBatch::Make(new_schema, num_rows(), columns);
```
Comment (pitrou) [inline cpp/src/arrow/record_batch.cc:274]: `std::move(columns)` will avoid a vector copy here.

**Review by pitrou — APPROVED** [review]
Looks good to me on the principle, just two comments if you want to polish this.

File: cpp/src/arrow/record_batch.cc (hunk 3 above)
Comment (thisisnic) [reply cpp/src/arrow/record_batch.cc:274]: Thanks - will try to make both of these changes as I'm new to C++ so useful for me to learn!

File: cpp/src/arrow/record_batch.cc (hunk 2 above)
Comment (thisisnic) [reply cpp/src/arrow/record_batch.cc]: Done!

File: cpp/src/arrow/record_batch.cc (hunk 3 above)
Comment (thisisnic) [reply cpp/src/arrow/record_batch.cc]: Done!

**Comment by pitrou** [comment]
Thanks for the update. I will wait for CI before merging.

**Comment by pitrou** [comment]
CI failures are unrelated. Thank you @thisisnic !

**Pull request closed without merging**